dist_run.py

```diff
@@ -388,88 +377,74 @@ def main(args):
-    elif pp_rank == 0:
-        src = dist.get_global_rank(pp_group, last_pp_group)
-
-    # Decoding
     num_tokens = 40
 
+    # Decoding
     with torch.no_grad():
         for step in range(num_tokens):
-            # first
-            if pp_rank == 0:
-                schedule.step(padded_sequence)
-                # only receive if not last step
-                if step < num_tokens - 1:
-                    dist.recv(
-                        x_recv,
-                        src,
-                        group=pp_group,
-                    )
-                    _update_padded_sequence(
-                        padded_sequence, x_recv, res, prompt_lengths
-                    )
-
-            # last
-            elif pp_rank == last_pp_group:
+            # Run data through pipeline
+            if pp_rank == first_pp_rank:
+                output = schedule.step(padded_sequence)
+            elif pp_rank == last_pp_rank:
                 output = schedule.step()
-                # need to decode the output
+            else:  # middle pp ranks
+                schedule.step()
+
+            # Decode the output
+            if pp_rank == last_pp_rank:
                 decode_results = _batch_decode_next_tokens(
                     output=output, prompt_lengths=prompt_lengths, tokenizer=tokenizer
                 )
                 if tp_rank == 0:
                     logger.info(
-                        f"\n\n{color.green} {'Prefill' if step == 0 else '* Decode *'} responses ====>>>> {color.blue} {decode_results=} \n{color.reset}"
+                        f"{color.green} {'Prefill' if step == 0 else '* Decode *'} "
+                        f"responses ====>>>> {color.blue} {decode_results=}{color.reset}"
                     )
-
-                next_token = torch.tensor([decode_results[0][0]], device=device)
+                # decode results returns both token_id (int) and token_str (readable), hence [0] and [1]
+                new_token = torch.tensor([decode_results[0][0]], device=device)
                 res.append(decode_results[0][1])
 
-                # increment prompt lengths for next token
-                for i in range(len(prompt_lengths)):
-                    prompt_lengths[i] += 1
-                    # logger.info(
-                    #     f"output review {prompt_lengths[i]=}, {padded_sequence[i, prompt_lengths[i]-1]=}"
-                    # )
-
-                # only send if not last step
-                if step < (num_tokens - 1):
-                    dist.send(
-                        next_token,
-                        dst,
-                        pp_group,
-                    )
+            # sendrecv between last and first ranks, only if:
+            # first_pp_rank != last_pp_rank.
+            if pp_rank == last_pp_rank and pp_rank != first_pp_rank:
+                dist.send(
+                    new_token,
+                    dst=dist.get_global_rank(pp_group, first_pp_rank),
```

Comment:
dst and src shouldn't be getting recreated every iter, since they don't change on a per iter basis. 
This is why I had moved them out of the loop previously. 
Not sure how expensive the dist.get_global_rank is but no need imo to be calling it over and over here.

Reply:
Makes sense. Will upload a PR to improve it. Thanks!